LOOP-1489: migrate charts to LoopKitUI

novalegra opened the pull request:
In order for the insulin model view to 'live' in LoopKitUI, most of the charts-related logic needs to move over as well. Apologies for this being on the larger side for a PR, though almost all of the additions are from moving files over.

Commit:
First pass

Commit:
Move LoopSettings into LoopKit

Commit:
Merge remote-tracking branch 'origin/dev' into LOOP-1484-CR

Commit:
Add informational view

Commit:
Add help screen to setting description

Commit:
Commenting

Commit:
Move CorrectionRangeOverridesEditor to LoopKit (& put editors in folder)

Commit:
Modify CorrectionRangeOverridesEditor for prescription flow

Commit:
Don't dismiss screen if in prescription flow

Commit:
Don't dismiss screen if we're in the flow

Commit:
Merge branch 'LOOP-1484-CR' into LOOP-1485-temp-cr

Commit:
Add temp correction range settings screen

Commit:
Don't display warning if displaying instructional content

Commit:
Don't display warning if displaying instructional content

Commit:
Merge branch 'LOOP-1484-CR' into LOOP-1485-temp-cr

Commit:
Make vars private

Commit:
Make title in-line

Commit:
Get changes

Commit:
Don't cut text off

novalegra committed:
Merge branch 'dev' into novalegra/LOOP-1484

Commit:
Merge branch 'novalegra/LOOP-1484' into LOOP-1485-temp-cr

Commit:
Add info screen

Commit:
Fix layout issues on iPhone SE

Commit:
Merge branch 'novalegra/LOOP-1485' into LOOP-1486-suspend

Commit:
Move suspend threshold editor to LoopKit and make it reusable

Commit:
Create basal rate info view

Commit:
Add basal rates editor

Commit:
Make button update in response to schedule changes

Commit:
Add placeholder saving mechanism

Commit:
Add setting to info page

Commit:
Make info button display correct info page

Commit:
Add delivery limits info screen

Commit:
Move DeliveryLimitsEditor to LoopKit

Commit:
Make DeliveryLimitsEditor configurable

Commit:
Enable save button if in flow

Commit:
Add insulin model info screen

Commit:
in progress changes

Commit:
wip changes

Commit:
Add notification observers back

Commit:
Make protocol

Commit:
Merge remote-tracking branch 'origin/dev' into migrating-insulin-model

Commit:
more wip

Commit:
Add view into LoopKit

Commit:
Rename to avoid confusion

Commit:
Merge branch 'dev' into migrating-insulin-model

Commit:
Remove button text as argument

Commit:
Make delegate public

Commit:
Merge remote-tracking branch 'origin/dev' into merging-working

Commit:
Pre-PR cleanup

novalegra commented on LoopKitUI/View Controllers/ChartsTableViewContainer.swift:
```diff
@@ -0,0 +1,189 @@
+import HealthKit
+import os.log
+
+public protocol ChartsTableViewControllerDelegate: class {
```
This protocol adds the needed logic for the charts to update when the glucose unit changes value

novalegra commented on LoopKitUI/View Controllers/ChartsTableViewContainer.swift:
```diff
@@ -0,0 +1,189 @@
+}
+
+/// Abstract class providing boilerplate setup for chart-based table view controllers
+open class ChartsTableViewContainer: UITableViewController, UIGestureRecognizerDelegate {
```
Open to different names for this

ps2 replied:
I would suggest keeping ChartsTableViewController for this, and renaming the inheriting class in Loop, to something like LoopChartsTableViewController

novalegra commented on LoopKitUI/Views/ChartTableViewCell.swift:
```diff
@@ -0,0 +1,62 @@
+        chartContentView.chartGenerator = nil
+    }
+
+    public func reloadChart() {
```
All of these functions were created because `IBOutlets` shouldn't be accessible outside of their defining module

Commit:
Move delivery limits to correct place

Commit:
Correctly move it over :)

Commit:
Move editor, not just reference

novalegra commented on LoopKitUI/Views/QuantityScheduleEditor.swift:
```diff
@@ -33,7 +33,6 @@ struct QuantityScheduleEditor<ActionAreaContent: View>: View {
     var guardrailWarning: (_ crossedThresholds: [SafetyClassification.Threshold]) -> ActionAreaContent
     var savingMechanism: SavingMechanism<DailyQuantitySchedule<Double>>
     var mode: PresentationMode
-    var buttonText: Text
```
This is stuff that should have been removed in my delivery limits PR but got overlooked

Commit:
Remove unused storyboard

Commit:
Add swiftcharts dependency to LoopKitUI

Commit:
Remove embedding

Commit:
WIP moving insulin model over

Commit:
Fix permissions issues

Commit:
Merge branch 'dev' into integrate-model

Commit:
Move Loopsettings back to Loop

Commit:
The parent should manage the BG unit updates

Commit:
Remove comment

Commit:
Remove logging from ChartsTableViewContainer

Commit:
Merge remote-tracking branch 'origin/dev' into merging

Commit:
Remove redundant files

novalegra commented on Extensions/HKUnit.swift:
```diff
@@ -41,4 +41,13 @@ extension HKUnit {
     }
+    
+    /// The smallest value expected to be visible on a chart
+    var chartableIncrement: Double {
```
Added because it's needed by the insulin model chart

Commit:
Remove diff

novalegra commented on LoopKitUI/ChartColorPalette.swift:
```diff
@@ -0,0 +1,30 @@
+//
+//  ChartColorPalette.swift
```
Moved because it's needed by insulin model chart

novalegra commented on LoopKitUI/Charts/GlucoseChart.swift:
```diff
@@ -0,0 +1,57 @@
+import SwiftCharts
+
+
+open class GlucoseChart {
```
Moved because it's needed by insulin model chart

novalegra commented on LoopKitUI/Extensions/ChartLineModel+LoopKitUI.swift:
```diff
@@ -0,0 +1,23 @@
+import SwiftCharts
+
+
+extension ChartLineModel {
```
Added because it's needed by insulin model chart

novalegra commented on LoopKitUI/Extensions/ChartSettings+LoopKitUI.swift:
```diff
@@ -0,0 +1,23 @@
+
+
+extension ChartSettings {
+    static var `default`: ChartSettings {
```
Added because it's needed by insulin model chart

novalegra commented on LoopKitUI/Extensions/DateFormatter.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+extension DateFormatter {
```
Added because it's needed by glucose chart, which is needed by insulin model chart

Commit:
Remove more logging from LoopKitUI

novalegra commented on LoopKitUI/Models/ChartAxisValueDoubleUnit.swift:
```diff
@@ -0,0 +1,31 @@
+//
+//  ChartAxisValueDoubleUnit.swift
```
Moved because it's needed by GlucoseChart, which is needed for the insulin model chart

novalegra commented on LoopKitUI/View Controllers/ChartsManager.swift:
```diff
@@ -0,0 +1,236 @@
+import SwiftCharts
+
+
+open class ChartsManager {
```
Moved because it's needed in ChartsTableViewContainer

novalegra commented on LoopKitUI/View Controllers/ChartsTableViewContainer.swift:
```diff
@@ -0,0 +1,151 @@
+
+    // MARK: - State
+    // This function should only be called from the main thread
+    public func unitPreferencesDidChange(to unit: HKUnit?) {
```
This was added by me to enable the parent view containing the chart to manage chart updates in the event of a glucose unit change

novalegra commented on LoopKitUI/Views/ChartContainerView.swift:
```diff
@@ -0,0 +1,51 @@
+//
+//  ChartContainerView.swift
```
Moved because it's needed by the insulin model chart

Commit:
Add Cartfile

Commit:
Update cartfile & build targets

Commit:
Copy frameworks

Commit:
Set up homebrew/carthage if needed

Commit:
Add Swiftcharts framework

Commit:
Remove SwiftCharts framework from LoopKit Example

Commit:
Don't embed

Commit:
Removed unneeded cartfile

Commit:
Merge remote-tracking branch 'origin/dev' into novalegra/LOOP-1489-migrate-charts

Commit:
Fix merge

Commit:
Merge fix

Commit:
More merge fixes

Commit:
Pull in Michael's changes

ps2 commented on LoopKitUI/Views/Settings Editors/copy-frameworks.sh:
```diff
@@ -0,0 +1,41 @@
+#!/bin/sh -e
```
Guessing this file was included accidentally.

novalegra replied:
Removed

ps2 commented on LoopKitUI/Extensions/UIColor+LoopKit.swift:
```diff
@@ -8,6 +8,22 @@
+
+    @nonobjc static let axisLineColor = UIColor.clear
+
+    @nonobjc static let doseTintColor = UIColor.systemOrange
```
This should not be hardcoded in a reusable chart system, but since we have separate planned work to setup a Loop level color palette, this is fine for now.

novalegra replied:
I pinged Nate about it as a head-up

Commit:
Edit circleci config

ps2 reviewed (changes requested):
Looks like a file was accidentally included that should be removed, but other than that just naming suggestions. Looks great!

Commit:
Remove unneeded script + update CircleCI

Commit:
Renaming (thanks for the feedback Pete!)

rickpasetto commented on .circleci/config.yml:
```diff
@@ -17,6 +17,31 @@ jobs:
       xcode: 11.5.0
     steps:
       - checkout
+      - run:
```
Why put this in the CircleCI config, and not in the Xcode project?

novalegra replied:
I put this in (and removed the Cartfile target) per a discussion with Pete about the options for integrating SwiftCharts. Basically we could use the already-existing SwiftCharts submodule in the workspace, or use Carthage with caching. Since we really only have the Carthage for the purpose of the CI builds, it was recommended that I remove the Cartfile target from the project and add those build steps to the CI build.

rickpasetto replied:
So how will DIY users build LoopKit after this?

ps2 replied:
Main DIY builds via Carthage through Loop, and for the DIY workspace build, we also include SwiftCharts project directly.

rickpasetto replied:
Is it okay that DIY Xcode builds from the `LoopKit` Xcode project won't work?

novalegra replied:
I have a similar question - based on how I _think_ Carthage works, wouldn't it be insufficient to have the Loop Carthage be pulling in SwiftCharts?

rickpasetto replied:
Why not remove SwiftCharts from the list of submodules? We're not changing it.  So let's just pull it in via Carthage, either in DIY or Tidepool builds.  Would that be possible, @ps2 ?

novalegra replied:
I think the concern is speed - Carthage is slower than the submodules, no?

rickpasetto replied:
Carthage (bootstrap) only needs to run once

ps2 replied:
I have several times modified SwiftCharts; it's a small project, and they've taken a number of DIY's changes.  It's open source, and building it as a first class member of our project is easy enough. And yes, even with caching, it does add significant time on the first build.

ps2 replied:
> Is it okay that DIY Xcode builds from the LoopKit Xcode project won't work?

Without running `carthage bootstrap` first? Yes. That's how several of the DIY dependencies work right now.  The main thing is that the non-workspace build of DIY Loop needs to be able to build everything as a zip, and this setup allows that, and also allows us to have speed and easy access to the SwiftCharts source.

ps2 replied:
> I have a similar question - based on how I think Carthage works, wouldn't it be insufficient to have the Loop Carthage be pulling in SwiftCharts?

Loop Cartfile references LoopKit, which makes Carthage fetch the Cartfile for LoopKit and resolve those dependencies too. So the Loop project, in xcode, when it runs the Cartfile target will end up building SwiftCharts, even if Loop's Cartfile didn't reference it.

rickpasetto replied:
I guess I'm still not convinced this is the right approach.  So as not to block @novalegra I will approve.

rickpasetto commented on LoopKit.xcodeproj/project.pbxproj:
```diff
@@ -3020,6 +3087,29 @@
+			inputFileListPaths = (
+			);
+			inputPaths = (
+				"$(BUILT_PRODUCTS_DIR)/SwiftCharts.framework/SwiftCharts",
```
I would have expected this input path to read

```
				"$(BUILT_PRODUCTS_DIR)/SwiftCharts.framework",
```

novalegra replied:
Doing so breaks the build 🤷‍♀️

rickpasetto replied:
Ok.  I'm still not sure why we need that separate script instead of just using `carthage copy-frameworks` directly, but I'll yield.

rickpasetto commented on LoopKit.xcodeproj/project.pbxproj:
```diff
@@ -4159,11 +4279,11 @@
 			buildSettings = {
 				APPLICATION_EXTENSION_API_ONLY = YES;
 				CLANG_ENABLE_MODULES = YES;
-				CODE_SIGN_IDENTITY = "";
```
I think you should revert this

rickpasetto commented on LoopKit.xcodeproj/project.pbxproj:
```diff
@@ -4159,11 +4279,11 @@
 				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "";
 				CODE_SIGN_STYLE = Automatic;
 				DEFINES_MODULE = YES;
-				DEVELOPMENT_TEAM = "";
```
I think you should revert this

rickpasetto commented on LoopKit.xcodeproj/project.pbxproj:
```diff
@@ -4184,11 +4304,11 @@
 			buildSettings = {
 				APPLICATION_EXTENSION_API_ONLY = YES;
 				CLANG_ENABLE_MODULES = YES;
-				CODE_SIGN_IDENTITY = "";
```
I think you should revert this

rickpasetto commented on LoopKit.xcodeproj/project.pbxproj:
```diff
@@ -4184,11 +4304,11 @@
 				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "";
 				CODE_SIGN_STYLE = Automatic;
 				DEFINES_MODULE = YES;
-				DEVELOPMENT_TEAM = "";
```
I think you should revert this

rickpasetto commented on LoopKit.xcodeproj/project.pbxproj:
```diff
@@ -4349,10 +4469,10 @@
 			buildSettings = {
 				APPLICATION_EXTENSION_API_ONLY = YES;
 				CLANG_ENABLE_MODULES = YES;
-				CODE_SIGN_IDENTITY = "";
```
I think you should revert this

rickpasetto commented on LoopKit.xcodeproj/project.pbxproj:
```diff
@@ -4349,10 +4469,10 @@
+				CODE_SIGN_IDENTITY = "Apple Development";
 				CODE_SIGN_STYLE = Automatic;
 				DEFINES_MODULE = YES;
-				DEVELOPMENT_TEAM = "";
```
I think you should revert this

rickpasetto commented on LoopKit.xcodeproj/project.pbxproj:
```diff
@@ -4373,10 +4493,10 @@
 			buildSettings = {
 				APPLICATION_EXTENSION_API_ONLY = YES;
 				CLANG_ENABLE_MODULES = YES;
-				CODE_SIGN_IDENTITY = "";
```
I think you should revert this

rickpasetto commented on LoopKit.xcodeproj/project.pbxproj:
```diff
@@ -4373,10 +4493,10 @@
+				CODE_SIGN_IDENTITY = "Apple Development";
 				CODE_SIGN_STYLE = Automatic;
 				DEFINES_MODULE = YES;
-				DEVELOPMENT_TEAM = "";
```
I think you should revert this

rickpasetto commented on LoopKitUI/Views/Settings Editors/InsulinModelSelectionView.swift:
```diff
@@ -0,0 +1,317 @@
+//
+//  InsulinModelSelection.swift
```
I'm confused...is this a duplicate file?

novalegra replied:
We love merge conflicts - removed

rickpasetto reviewed (changes requested):
Wow, this was a lot of work!  I admit I didn't scan through every moved file... can you maybe annotate where you had to make changes so it is easier to review?
That said, there are still some changes that need to be made to this before it can go in.

Commit:
Remove duplicate file

Commit:
Fix copying

Commit:
Remove team refs

novalegra commented on LoopKitUI/View Controllers/ChartsTableViewController.swift:
```diff
@@ -0,0 +1,151 @@
+/// Abstract class providing boilerplate setup for chart-based table view controllers
+open class ChartsTableViewController: UITableViewController, UIGestureRecognizerDelegate {
+
+    public var preferredGlucoseUnit: HKUnit?
```
This was added by me to shift the responsibility of unit updates to the view containing the chart

novalegra commented on LoopKitUI/View Controllers/ChartsTableViewController.swift:
```diff
@@ -0,0 +1,151 @@
+    open override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        if let unit = preferredGlucoseUnit {
+            self.charts.setGlucoseUnit(unit)
+        }
```
Added by me

novalegra commented on LoopKitUI/View Controllers/ChartsTableViewController.swift:
```diff
@@ -0,0 +1,151 @@
+
+    // MARK: - State
+    // This function should only be called from the main thread
+    public func unitPreferencesDidChange(to unit: HKUnit?) {
+        if let unit = unit {
+            self.charts.setGlucoseUnit(unit)
+            self.glucoseUnitDidChange()
+        }
+        self.reloadData()
+    }
```
This is a slightly modified version of what was previously present (essentially, it's now a public function that is called by the parent, instead of an `@objc` function that's called by an observer)

Commit:
Merge branch 'dev' into novalegra/LOOP-1489-migrate-charts

novalegra commented:
@rickpasetto @ps2 I addressed your feedback and noted the (very limited places) where I made changes relative to the files that were moved in. Most of my work was adding `public`s and `open`s to the right places 😄

Commit:
Add todo

Commit:
Update todo

mpangburn commented on LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift:
```diff
@@ -0,0 +1,317 @@
+                        )
+                        .frame(height: 170)
+
+                        InsulinModelSettingsRow(
```
Looks like somewhere along the way there may have been a merge issue here, I believe @rickpasetto refactored this checkmark row component, and it was renamed

novalegra replied:
I just re-copy-pasted in. If there are any other screens that I moved from Loop->LoopKit that have been modified within the last week, I may not have the latest copy of the changes. I think this is basically the only one, though, aside from the color changes @nhamming is making

mpangburn commented on LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift:
```diff
@@ -0,0 +1,317 @@
+                    }
+
+                    if supportedModelSettings.walshModelEnabled {
+                        DurationBasedInsulinModelSettingsRow(
```
(same with the other insulin model rows and this duration-based one)

mpangburn commented on LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift:
```diff
@@ -0,0 +1,317 @@
+    }
+}
+
+fileprivate struct InsulinModelSettingsRow: View {
```
and then these fileprivate structs ended up being removed in favor of the extracted component

mpangburn reviewed (commented):
I think there might've been a merge issue somewhere along the way

Commit:
Pull latest changes

Commit:
Merge remote-tracking branch 'origin/dev' into novalegra/LOOP-1489-migrate-charts

Commit:
Remove duplicate extension

Commit:
Add missing "publics"

Commit:
Fix merge issue

rickpasetto reviewed (approved):
I still think the carthage scripts should go in the Xcode project and not the CircleCI script, but let's get this in so we can move forward.

mpangburn reviewed (approved):
Thanks for the heavy lift moving all this down!

Commit:
Merge remote-tracking branch 'origin/dev' into novalegra/LOOP-1489-migrate-charts

ps2 reviewed (approved):
LGTM!
